lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1606,12 +1625,13 @@ int swift::performFrontend(ArrayRef<const char *> Args,
   // arguments are generated by the driver, not directly by a user. The driver
   // is responsible for emitting diagnostics for its own errors. See SR-2683
   // for details.
+  //
+  // FIXME: Do we need one SerializedConsumer per primary? Owned by the
+  // SourceFile?
   std::unique_ptr<DiagnosticConsumer> SerializedConsumer;
   {
-    const std::string &SerializedDiagnosticsPath =
-        Invocation.getFrontendOptions()
-            .InputsAndOutputs.supplementaryOutputs()
-            .SerializedDiagnosticsPath;
+    const std::string SerializedDiagnosticsPath =
```

Comment:
Why make this a `std::string` at all? (here and elsewhere). You know the Invocation is alive in the current stack frame, so everything it owns is alive too.

Reply:
And if it's because we need the path to be a C string, that's probably a reason to just change all the path-returning methods to use `const std::string &` instead of `StringRef`. That's an exception I forgot to mention that Graydon already ran into on the Driver side.

Reply:
Changed it to a StringRef. (But in general StringRefs give me a queasy feeling because one has to know a lot to guarantee safety.)

As far as the C string issue, which is not here, but may be elsewhere, are you saying that all the path-returning methods should be changed to `const std::string &`?

Reply:
`const std::string &` provides a guarantee that StringRef doesn't, which is null-termination, at the cost of not being able to use other string types (like llvm::SmallString). If that guarantee would save on copies elsewhere, it might be worth it.

It's true that StringRef isn't safe-by-design and std::string is, but copies add up. I view this as the same sort of eating-your-vegetables performance as bitfield packing: it's not that any individual copy will make things worse, but that just means that if you don't avoid copies by default things will be slower and the profile won't show any hotspots.

Reply:
I think of two issues here:

1. `StringRef` vs `const std::string &`: It seems to me that since we are storing strings with null termination, I might as well have the routines in question return `const std::string &` as you suggest. I'll make the changes.

2. `StringRef` (or `std::string const &`) vs copying a `std::string`. There are two sub-issues here for me: the cost and the benefits:

2a: The cost of copying: (I'm going to be very very approximate here.) Suppose the average path string is 100 characters, and that it costs, say 1000 clocks to copy it. Then on a 3GHz machine, that's 1 ns per copy, if I've the math right. Suppose we're talking about 100 of these strings, that's 100 ns to copy the whole set. Now this path copying isn't (as far as I know) done over-and-over, so I'll assume the whole set is copied 10 times. (Am I wrong here?) (These aren't identifier names that might be looked up every time they are used in a program, but just file names which are mainly used to open output files.) So the whole cost of copying these names is 1 µs. If I have this right, we're talking 1 µs to copy these path names instead of passing references. That's the cost.

2b. What's the benefit? (Or, conversely, what's the cost of passing refs?) I don't think the analogy to bitfield packing holds because I can isolate bitfield packing to a single, simple class and thus reason fairly locally about the correctness of the optimization. OTOH, in order to be sure  ref-passing code is correct, I have to first follow a call-chain down to a leaf function so I can see where the actual string is allocated, then I have to follow the structure chain up to find the outermost enclosing structure, then I have to figure out where that structure might be freed. The reasoning required to understand ref-passing code seems to be much much much more harder, time-consuming, and error-prone than the reasoning required to verify bit-field packing.

So, my conclusion is that the tradeoff in this case leans quite heavily towards copying strings and away from passing references. I fear that, even if I get all the tests to pass for a ref-passing version, I will have creating lurking bugs. Or, someday, someone else may evolve the code in a way as to create lurking bugs. If one such is created and is reproducible, than a sanitizer can help find and fix them, but that seems like a high price to pay.

Please let me know what you think. I may well be missing some important fact about the compiler, or C++, or something else. If so, it would help me do better work for the team to correct my misunderstanding.

For now, I will switch over to the string refs.

Thanks!